internal/work/dispatcher.go

```diff
@@ -0,0 +1,335 @@
+package work
+
+import (
+	"fmt"
+	"io"
+	"net/url"
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"git.sr.ht/~spc/go-log"
+	"github.com/godbus/dbus/v5"
+	"github.com/godbus/dbus/v5/introspect"
+	"github.com/redhatinsights/yggdrasil"
+	internalhttp "github.com/redhatinsights/yggdrasil/internal/http"
+	"github.com/redhatinsights/yggdrasil/internal/sync"
+	"github.com/redhatinsights/yggdrasil/ipc"
+)
+
+type DispatcherEvent uint
+
+const (
+	DispatcherEventReceivedDisconnect   DispatcherEvent = 1
+	DispatcherEventUnexpectedDisconnect DispatcherEvent = 2
+	DispatcherEventConnectionRestored   DispatcherEvent = 3
+)
+
+// Dispatcher implements the com.redhat.yggdrasil.Dispatcher1 D-Bus interface
+// and is suitable to be exported onto a bus.
+//
+// Dispatcher receives values on its 'inbound' channel and sends them via D-Bus
+// to the destination worker. It sends values on the 'outbound' channel to relay
+// data received from workers to a remote address.
+type Dispatcher struct {
+	HTTPClient  *internalhttp.Client
+	conn        *dbus.Conn
+	features    sync.RWMutexMap[map[string]string]
+	senders     sync.RWMutexMap[string]
+	Dispatchers chan map[string]map[string]string
+	Inbound     chan yggdrasil.Data
+	Outbound    chan struct {
+		Data yggdrasil.Data
+		Resp chan struct {
+			Code     int
+			Metadata map[string]string
+			Data     []byte
+		}
+	}
+}
+
+func NewDispatcher(client *internalhttp.Client) *Dispatcher {
+	return &Dispatcher{
+		HTTPClient:  client,
+		features:    sync.RWMutexMap[map[string]string]{},
+		senders:     sync.RWMutexMap[string]{},
+		Dispatchers: make(chan map[string]map[string]string),
+		Inbound:     make(chan yggdrasil.Data),
+		Outbound: make(chan struct {
+			Data yggdrasil.Data
+			Resp chan struct {
+				Code     int
+				Metadata map[string]string
+				Data     []byte
+			}
+		}),
+	}
+}
+
+// Connect examines the environment for a private D-Bus address and connects to
+// it if found. Otherwise, it connects to the system bus.
+func (d *Dispatcher) Connect() error {
+	var conn *dbus.Conn
+	var err error
+	if os.Getenv("DBUS_SESSION_BUS_ADDRESS") != "" {
```

Comment:
Shouldn't this be part of the connect args? Getting ENV variable in the package It's something that I'm not big fan.

Reply:
Sort of. `DBUS_SESSION_BUS_ADDRESS` is a well-known environment variable name defined by the [dbus-daemon](https://dbus.freedesktop.org/doc/dbus-daemon.1.html) specification. Because it's set by the broker, I think it's safe to assume to be in the environment of the process. I did discover `DBUS_STARTER_BUS_TYPE` while rereading the man page just now, and I think I can improve this logic to use that variable.